src/server/pfs/server/driver_fsck.go

```diff
@@ -134,163 +134,231 @@ func (e ErrMissingBranchHead) Error() string {
+
+	if bi.Head == nil {
+		if err := onError(ErrMissingBranchHead{
+			Branch: bi.Branch,
 		}); err != nil {
-			return errors.EnsureStack(err)
+			return err
+		}
+	} else {
+		// we expect the branch's provenance to equal the HEAD commit's provenance
+		// i.e branch.Provenance contains the branch provBranch and
+		// provBranch.Head != nil implies branch.Head.Provenance contains
+		// provBranch.Head
+		for _, provBranch := range bi.Provenance {
+			provBranchInfo, ok := branchInfos[pfsdb.BranchKey(provBranch)]
+			if !ok {
+				if err := onError(ErrBranchInfoNotFound{Branch: provBranch}); err != nil {
+					return err
+				}
+				continue
+			}
+			if provBranchInfo.Head != nil {
+				// in this case, the headCommit Provenance should contain provBranch.Head
+				if _, ok := commitInfos[pfsdb.CommitKey(bi.Head)]; !ok {
+					if err := onError(ErrCommitInfoNotFound{
+						Location: "head commit provenance (=>)",
+						Commit:   bi.Head,
+					}); err != nil {
+						return err
+					}
+				}
+			}
 		}
-		branchInfo := &pfs.BranchInfo{}
-		err := d.branches.ReadOnly(ctx).GetByIndex(pfsdb.BranchesRepoIndex, pfsdb.RepoKey(repoInfo.Repo), branchInfo, col.DefaultOptions(), func(string) error {
-			branchInfos[pfsdb.BranchKey(branchInfo.Branch)] = proto.Clone(branchInfo).(*pfs.BranchInfo)
-			return nil
-		})
-		return errors.EnsureStack(err)
-	}); err != nil {
-		return errors.EnsureStack(err)
 	}
 
+	return nil
+}
+
+
+func FsckBranches(branchInfos map[string]*pfs.BranchInfo, commitInfos map[string]*pfs.CommitInfo, onError func(error) error) error {
+
 	// for each branch
 	for _, bi := range branchInfos {
-		// we expect the branch's provenance to equal the union of the provenances of the branch's direct provenances
-		// i.e. union(branch, branch.Provenance) = union(branch, branch.DirectProvenance, branch.DirectProvenance.Provenance)
-		direct := bi.DirectProvenance
-		union := []*pfs.Branch{bi.Branch}
-		for _, directProvenance := range direct {
-			directProvenanceInfo := branchInfos[pfsdb.BranchKey(directProvenance)]
-			union = append(union, directProvenance)
-			if directProvenanceInfo != nil {
-				union = append(union, directProvenanceInfo.Provenance...)
-			}
+
+		if err := checkBranchProvenances(bi, branchInfos, onError); err != nil {
+			return err
+		}
+
+		// every provenant branch should have this branch in its subvenance
+		if err := checkBranchSubvenances(bi, branchInfos, onError); err != nil {
+			return err
+		}
+
+		if err:= checkMissingHead(bi, branchInfos, commitInfos, onError); err != nil {
+			return err
 		}
 
-		if !equalBranches(append(bi.Provenance, bi.Branch), union) {
-			if err := onError(ErrBranchProvenanceTransitivity{
-				BranchInfo:     bi,
-				FullProvenance: union,
+	}
+
+	return nil
+}
+
+func CheckParentCommit(commitInfo *pfs.CommitInfo, commitInfos map[string]*pfs.CommitInfo, onError func(error) error) error {
```

Comment:
nit: could you camel case these internal functions? upper-casing them exposes them as "public" in the go package which is not clear that we want right now